### PR TITLE
fix: do not reshard parameters twice

### PR DESCRIPTION
**Pull request by 0x804d8000**
This PR fixes potential double resharding of parameters that both:

1. requires no gradient and,
2. were used more than once during forward pass.

[`_register_post_backward_hook`](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_runtime_utils.py#L1415) handles the case correctly, this PR does the same for `_register_post_backward_reshard_only_hook`.

### CODE REVIEW

**Comment by awgu** [comment]
Thanks @0x804d8000! Would you be interested in adding a unit test?

https://github.com/pytorch/pytorch/blob/main/test/distributed/fsdp/test_fsdp_fine_tune.py

**Comment by 0x804d8000** [comment]
Sure. Added a test that fails on main branch and passes with the PR applied.

**Comment by 0x804d8000** [comment]
By the way, initially I came up with a module as:

```python
def _init_my_module(self) -> nn.Module:
    torch.manual_seed(42)
    # The same layer is traversed for multiple times.
    layers = [nn.Linear(5, 5, device="cuda"), nn.ReLU()] * self.NUM_LINEARS
    seq = nn.Sequential(*layers)
    self._set_seq_module_requires_grad(seq, False)
    return seq
```

Not sure if such a module is meant to be supported by FSDP with the following wrap policy:

```python
seq = self._init_my_module()
policy = ModuleWrapPolicy({nn.Linear})
seq = FSDP(
    seq,
    auto_wrap_policy=policy,
    sharding_strategy=sharding_strategy,
    use_orig_params=use_orig_params,
)
print(seq)
```

But if it does, the wrapped module looks weird to me, only the first `nn.Linear` is wrapped:

```python
FullyShardedDataParallel(
  (_fsdp_wrapped_module): Sequential(
    (0): FullyShardedDataParallel(
      (_fsdp_wrapped_module): Linear(in_features=5, out_features=5, bias=True)
    )
    (1): ReLU()
    (2): Linear(in_features=5, out_features=5, bias=True)
    (3): ReLU()
    (4): Linear(in_features=5, out_features=5, bias=True)
    (5): ReLU()
    (6): Linear(in_features=5, out_features=5, bias=True)
    (7): ReLU()
    (8): Linear(in_features=5, out_features=5, bias=True)
    (9): ReLU()
    (10): Linear(in_features=5, out_features=5, bias=True)
    (11): ReLU()
  )
)
```

I haven't come up with a fix for this case, but I'd like to bring it up here so people could discuss it.

**Comment by 0x804d8000** [comment]
Would anyone please approve and trigger CI, so I can fix any possible CI failure? Thanks in advance.

File: test/distributed/fsdp/test_fsdp_fine_tune.py [hunk 1]
```diff
@@ -151,6 +151,68 @@ def _assert_post_backward_reshard_count(step_idx, num_steps):
...
+            auto_wrap_policy=policy,
+            sharding_strategy=sharding_strategy,
+            use_orig_params=use_orig_params,
+        )
+        inp = torch.randn((8, 5), device="cuda", requires_grad=inp_requires_grad)
+        output = seq(inp)
+        output.sum().backward()
+
+        # No assertion here. The case exiting normally confirms that we do shard
+        # / unshard correctly.
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_fine_tune.py:214]: Thanks for the great unit test!

I think it would be better if we checked for numeric correctness too. I am not sure if DDP supports this case, but if so, then one option is to compare against DDP for numerics.

If DDP does not support this, then we can also compare against a local model, where the local model processes a global batch with batch size equal to `self.world_size * local_batch_size`, while the FSDP model processes a batch with size `local_batch_size`. Then, we can manually divide the local model's gradients by `self.world_size`, and we should expect numeric parity, which can be checked by running a training loop and comparing the forward outputs (or loss).

Comment (0x804d8000) [reply test/distributed/fsdp/test_fsdp_fine_tune.py]: Thanks for the feedback.

I reworked the test by referring to the implementation of `test_parity_with_ddp`. Now the test compares the loss to DDP's loss for correctness.

**Review by awgu — APPROVED** [review]
Looks great! Thanks for this!
